Source/FakeItEasy/Core/EventHandlerArgumentProviderMap.cs

```diff
@@ -1,29 +1,24 @@
 ﻿namespace FakeItEasy.Core
 {
     using System;
+    using System.Collections.Concurrent;
     using System.Collections.Generic;
 
     internal class EventHandlerArgumentProviderMap
     {
-        private readonly Dictionary<Delegate, IEventRaiserArguments> map
-            = new Dictionary<Delegate, IEventRaiserArguments>(new EventRaiserDelegateComparer());
+        private readonly ConcurrentDictionary<Delegate, IEventRaiserArguments> map
+            = new ConcurrentDictionary<Delegate, IEventRaiserArguments>(new EventRaiserDelegateComparer());
 
         public void AddArgumentProvider(Delegate eventHandler, IEventRaiserArguments argumentProvider)
         {
-            lock (this.map)
-            {
-                this.map[eventHandler] = argumentProvider;
-            }
+            this.map[eventHandler] = argumentProvider;
         }
 
         public IEventRaiserArguments TakeArgumentProviderFor(Delegate eventHandler)
         {
-            lock (this.map)
-            {
-                var provider = this.map[eventHandler];
-                this.map.Remove(eventHandler);
-                return provider;
-            }
+            IEventRaiserArguments provider;
+            this.map.TryRemove(eventHandler, out provider);
+            return provider;

```

Comment:
I'm not entirely sure how to answer this. The questions sound slightly leading, which makes me think you're trying to guide me to a realization about a deficiency in the type "in isolation". I fear I'm not getting it, but let's see if I can work myself up to it. Let's see where we end up.

I agree that the type should represent a sensible encapsulation and convey the right semantics. As such, I'd been satisfied with the type as it is. Setting aside the point that the signatures have not changed, so they convey the same semantics that they always did, we have a unit whose job is to provide a lookup table from event handlers to objects that can provide arguments that should be passed to an event handler. To support that, it offers the ability to add a `<hander, argument provider>` pair, to fetch and remove one, and also to check if there is such a record for a given handler.
This seemed to me a cohesive set of operations, working around a single responsibility.
However, there's nothing inherent in the signature of `TakeArgumentProviderFor` that indicates one way or another the behaviour when there's no matching argument provider for a supplied event handler, which could be considered a problem. We could change the signature to something like
`public bool TryTakeArgumentProviderFor(Delegate eventHandler, out IEventRaiserArguments argumentProvider)` in order to more clearly signal to the clients what's going on.
I don't hate it. I kind of like it, but will need to gauge how this affects the rest of the class and its clients

1. Given such a change, I think the need for `Contains` is greatly diminished, since production code only cares about whether an item is the map when it wants to use the arguments. As such, `EventRule.HandleEventCall` could be rewritten to attempt to take the arguments, and  on failure, add an event listener instead of firing one. A positive change, eliminating what I'd initially thought might've been your confusion about whether we'd ever try to take something from the map without it being there. Let's accept that.
    1. If `Contains` goes, `RaiseTests.Should_not_leak_handlers_when_raising` would have to change. I think it could use `TryTakeArgumentProviderFor` instead.
2. One may then notice that the signatures of the remaining methods on `EventHandlerArgumentProviderMap` match methods on `ConcurrentDictionary<Delegate, IEventRaiserArguments>` (One could even argue that `TryTakeArgumentProviderFor` could've been named `TryRemoveArgumentProviderFor`, as it more closely resembles the `TryRemove` precedent set by ConcurrentDictionary. Me, I prefer the `Take` in the name. I think it more clearly expresses "remove and return".),  and wonder about collapsing the class entirely. I'd be inclined against it myself, for reasons both aesthetic and practical:
    1. We'd have to decide where to hang the `EventRaiserDelegateComparer`. I think it fits very nicely nested inside its parent.
    2. I prefer to have a small, single-purpose class exposing its API to the clients, rather than the sprawling `ConcurrentDictionary` where we're supposed to limit our actions
    3. I think if we collapsed the type, we'd either have to start finding the instance using something like `ServiceLocator.Current.Resolve<ConcurrentDictionary<Delegate, IEventRaiserArguments>>()` or agument `ServiceLocator` to look items up by name or some such. Neither particularly appeals to me.

So, I've convinced myself that we may be better off with something like:

```csharp
internal class EventHandlerArgumentProviderMap
{
    public void AddArgumentProvider(Delegate eventHandler, IEventRaiserArgumentProvider argumentProvider)
    public bool TryTakeArgumentProviderFor(Delegate eventHandler, out IEventRaiserArgumentProvider argumentProvider)
}
```

(I sneaked a type rename in there:  `IEventRaiserArguments`  → `IEventRaiserArgumentProvider`, since the former made it sound like we had arguments, when really what we have is a thing that can make arguments for us.)

Has that at all ended up where you'd thought we might, or at least in pleasing location? Or have I misinterpreted your intent and further muddied the waters?